src/middlewared/middlewared/plugins/acme_protocol.py

```diff
@@ -0,0 +1,485 @@
+                'Please agree to the terms of service'
+            )
+
+        # For now we assume that only root is responsible for certs issued under ACME protocol
+        email = (self.middleware.call_sync('user.query', [['uid', '=', 0]]))[0]['email']
+        if not email:
+            raise CallError(
+                'Please specify root email address which will be used with the ACME server'
+            )
+
+        if self.middleware.call_sync(
+                'acme.registration.query', [['directory', '=', data['acme_directory_uri']]]
+        ):
+            verrors.add(
+                'acme_registration_create.acme_directory_uri',
+                'A registration with the specified directory uri already exists'
+            )
+
+        if verrors:
+            raise verrors
+
+        key = jose.JWKRSA(key=rsa.generate_private_key(
+            public_exponent=data['JWK_create']['public_exponent'],
+            key_size=data['JWK_create']['key_size'],
+            backend=default_backend()
+        ))
+        acme_client = client.ClientV2(directory, client.ClientNetwork(key))
+        register = acme_client.new_account(
+            messages.NewRegistration.from_data(
+                email=email,
+                terms_of_service_agreed=True
+            )
+        )
+        # We have registered with the acme server
+
+        # Save registration object
+        registration_id = self.middleware.call_sync(
+            'datastore.insert',
+            self._config.datastore,
+            {
+                'uri': register.uri,
+                'tos': register.terms_of_service,
+                'new_account_uri': directory.newAccount,
+                'new_nonce_uri': directory.newNonce,
+                'new_order_uri': directory.newOrder,
+                'revoke_cert_uri': directory.revokeCert,
+                'directory': data['acme_directory_uri']
+            }
+        )
+
+        # Save registration body
+        self.middleware.call_sync(
+            'datastore.insert',
+            'system.acmeregistrationbody',
+            {
+                'contact': register.body.contact[0],
+                'status': register.body.status,
+                'key': key.json_dumps(),
+                'acme': registration_id
+            }
+        )
+
+        return self.middleware.call_sync(f'{self._config.namespace}._get_instance', registration_id)
+
+
+class DNSAuthenticatorService(CRUDService):
+
+    class Config:
+        namespace = 'acme.dns.authenticator'
+        datastore = 'system.acmednsauthenticator'
+
+    def __init__(self, *args, **kwargs):
+        super(DNSAuthenticatorService, self).__init__(*args, **kwargs)
+        self.schemas = DNSAuthenticatorService.initialize_authenticator_schemas()
+
+    @accepts()
+    def authenticator_schemas(self):
+
+        """
+
+        Get all DNS providers we support for ACME DNS Challenge and the respective attributes required for connecting
```

Comment:
`Get all DNS providers`

These are schemas, no? Wouldnt `Get the schemas for all DNS providers...` make more sense?